rfcs/0010-eaglesong/0010-eaglesong.md

```diff
@@ -0,0 +1,335 @@
+---
+Number: 0010
+Category: Standards Track
+Status: Proposal
+Author: Alan Szepieniec
+Organization: Nervos Foundation
+Created: 2019-07-18
+---
+
+# Eaglesong (Proof-of-Work Function for Nervos CKB)
+
+This documents specifies the Eaglesong hash function as it is to be used in the context of Nervos CKB proof-of-work.
+
+ * [Notation](#Notation)
+ * [Design Strategies](#Desgin-Strategies)
+   * [Sponge Construction](#Sponge-Construction)
+   * [ARX](#ARX)
+ * [Round Function](#Round-Function)
+   * [Bit Matrix](#Bit-Matrix)
+   * [Circulant Multiplication](#Circulant-Multiplication)
+   * [Injection of Constants](#Injection-of-Constants)
+   * [Addition-Rotation-Addition](#Addition-Rotation-Addition)
+   * [Full Permutation](#Full-Permutation)
+ * [Eaglesong Hash](#Eaglesong-Hash)
+ * [Reference Implementations](#Reference-Implementations)
+
+
+<a name="Notation"></a>
+## Notation
+In the pseudocode snippets below the following notation is used:
+
+ - `//` -- denotes a comment
+ - `||` -- denotes concatenation (of bit strings)
+ - `length` -- returns the length (in bits) of the argument
+ - `%` -- modulo operator, computes the remainder of left-hand-side argument after division by the right-hand-side
+ - `zeros` -- produces a string of zero-bits of the specified length
+ - `xor` -- denotes the exclusive-or operation of equal-length bit strings
+ - `[]` -- python-style array indexing, with indexation starting (as it should) from zero; and when an upper-bound is included the range until but not including that upper bound is denoted; when not juxtaposed to an array, the expression `[a:b]` denotes the list of integers `{a, a+1, ..., b-1}`.
+
+<a name="Design-Strategies"></a>
+# Design Strategies
+
+<a name="Sponge-Construction"></a>
+## Sponge Construction
+
+Eaglesong is a [sponge construction](https://en.wikipedia.org/wiki/Sponge_function). This means that what is specified is a permutation `F` along with three variables:
+
+ - `r`, the rate : a positive integer;
+ - `c`, the capacity : another positive integer;
+ - `d`, the delimiter : a byte fixed to a nonzero value.
+
+The permutation maps `r+c` bits to `r+c` bits. After appending the delimiter byte to variable-length input to the hash function, this input is chunked into pieces `chunk[i]`, each of `r` bits long with thhe last one padded with zeros as necessary. The sponge construction defines a state, which consists of `r+c` bits initialized to zeros. The absorbing phase iterates over all chunks and xors the current chunk into the top `r` bits of the state before applying the permutation. In the squeezing phase, the output buffer is initialized to the empty string. Until this buffer is of the right size, the permutation is applied to the state and the top `r` bits are read out and appended to the output buffer. If the output buffer is too long, it is truncated. The next pseudocode illustrates this operation.
```

Comment:
```suggestion
The permutation maps `r+c` bits to `r+c` bits. After appending the delimiter byte to variable-length input to the hash function, this input is chunked into pieces `chunk[i]`, each of `r` bits along with the last one padded with zeros as necessary. The sponge construction defines a state, which consists of `r+c` bits initialized to zeros. The absorbing phase iterates over all chunks and xors the current chunk into the top `r` bits of the state before applying the permutation. In the squeezing phase, the output buffer is initialized to the empty string. Until this buffer is of the right size, the permutation is applied to the state and the top `r` bits are read out and appended to the output buffer. If the output buffer is too long, it is truncated. The next pseudocode illustrates this operation.
```

Reply:
Not sure what the suggested modification is here (line 52).

Reply:
it looks like github diff function is broken... just typo fix:
```diff
- each of `r` bits long with thhe last one padded with zeros as necessary.
+ each of `r` bits along with the last one padded with zeros as necessary. 
```

Reply:
Got it, thanks!